PSET integration tests

RCasatta opened the pull request:
~~on top of #100, see only 488cdacf7e1eff74fd90b1f7517f33033a619ce2 (MERGED)~~

~~on top of #109 (MERGED)~~

I would like to introduce some PSET integration testing against elements core.
To do so in a easy/maintainable/isolated way, I added the elementsd crate as a dev dependency.
This crate downloads the node binary if a feature is specified (or allow to specify the executable via env var)

I would like to gather early feedback on this approach before implementing the other cases: blinded, issuance, reissueance, pegin.

sanket1729 commented:
Strong Concept ACK. We do something similar in rust-miniscript. And I would also like to this for elements-miniscript

RCasatta commented:
Improved a bit, added blinded transaction, I would like to add also issuance, reissuance and pegin.

Not sure what's happening with CI and rust toolchain 1.29, any help is appreciated.

RCasatta commented on Cargo.toml:
```diff
@@ -30,17 +31,21 @@ slip21 = "0.2.0"
 
+# This should be an optional dev-dependency (only needed for integration tests),
+# but dev-dependency cannot be optional, and without optionality older toolchain try to compile it and fails
+elementsd = { version="0.3.0", features=["0_21_0","bitcoind_22_0"], optional = true }
```
I would like this to be dev-dependency but as written in the comment I am afraid it cannot be...

RCasatta replied:
It doesn't, there were some efforts to support it here  https://github.com/RCasatta/bitcoind/pull/31 but in the end, I thought it's fine for integration tests to run on stable

sanket1729 replied:
Agreed, it's fine for the integration tests to only be on stable

RCasatta commented on tests/pset.rs:
```diff
@@ -0,0 +1,335 @@
+        let psbt_base64 = elementsd.wallet_process_psbt(&psbt_base64);
+        assert_eq!(elementsd.expected_next(&psbt_base64), "extractor");
+
+        // TODO this fails because elements decodepsbt is not printing TxOut::asset (PSET_IN_WITNESS_UTXO)
```
I am using `fn psbt_rtt` to verify compatibility between pset implementations rust-elements and elements, but this requires `decodepsbt` to print everything in the upset and I think it's not the case as written in the above comment, maybe @stevenroose could confirm/deny

I cannot compare binaries because pair fields aren't ordered the same

sanket1729 replied:
This is tricky. I don't see any solution. Also, in general, I don't like the idea of testing using `decode`RPCs. If I recall correctly even decoderawtransaction does not print all details. But I don't see any other way for doing this :(

RCasatta replied:
Implementing `cmp::Eq` on PSET other than being a mess would not accomplish what we are trying to achieve here (PSET implementations compatibility). Suppose for example rust-elements deserialize is skipping a field in the PSET, you wouldn't notice. 

Since encoded binary comparison is excluded because the standard is not enforcing field ordering https://github.com/ElementsProject/elements/issues/1059 I can't think of another way to verify compatibility other than relying on `decodepsbt` printing all fields.

sanket1729 replied:
Unfortunately, I think we just need to comment this test. Raise an issue in upstream elements to add this, and add a remainder issue in rust-elements to re-add these once upstream elements fixes the bug.

RCasatta commented on tests/pset.rs:
```diff
@@ -49,6 +54,51 @@ mod tests {
+        let asset_id = AssetId::from_entropy(entropy.clone());
+        assert_eq!(
+            asset_id.to_string(),
+            "78b4920005fa0156ae3779129338bc2707e4d07cf8a0c2593583e3c1da3bb58c"
```
Not sure what's best here:
1) remove the asserts
2) add a comment about the possible change in upcoming elements release
3) I wanted to add something like `if cfg!(all(feature = "elementsd/0_21_0", feature = "elementsd/0_18_1_12"))`, but it looks rust doesn't support this unless we add features like `elementsd_0_21_0 = ["elementsd/0_21_0"]` which is probably not worthy for this, but it may be useful to have for other integration tests

sanket1729 replied:
We are (indirectly) asserting that the utxo (which is the first one obtained from `listunspent`) is a specific utxo with a fixed hard-coded outpoint. That might change in elements release because the order of coins in listunspent is updated (I am not sure if the ordering is guaranteed to be the same across multiple calls of elementsd of the same version). 

I am okay adding a comment here if you think there is some value in asserting the fixed asset_id (partly because I am curious to see if this ever breaks :) ). But if this ever breaks in the future, I think we should just remove it.

RCasatta replied:
I just removed the asserts in 682eb8a5bc972e28ccea8997ec582aa2d28769a1, they weren't really asserting something in the scope of this PR

RCasatta commented:
I tested with rust 1.36 and [it works](https://github.com/RCasatta/rust-elements/runs/4468744517?check_suite_focus=true) , I don't see a way to make it compatible with 1.29.

My understanding is that there is consensus to go with 1.36 in the rust-bitcoin community, so I guess it's fine also here to increase MSRV, and this can be merged then

stevenroose commented:
Can you rebase on master? @sanket1729 can you also review and ack?

stevenroose commented on Cargo.toml:
```diff
@@ -30,17 +31,21 @@ slip21 = "0.2.0"
 ryu = "<1.0.5"
 bincode = "1.3"
-
+base64 = "0.13.0"
```
It's ok as a dev-dep, but this crate has historically broken MSRV in a minor release. That's why I created [`base64-compat`](https://crates.io/crates/base64-compat) at the time..

RCasatta replied:
FYI In the past I suggested the base64 rust-bitcoin feature https://github.com/Kixunil/loptos/issues/1 but in that case it was refused because base64 already in deps

RCasatta commented:
Rebased

stevenroose commented on tests/pset.rs:
```diff
@@ -0,0 +1,335 @@
+            }
+            b_bytes[i] = b_bytes[i].wrapping_sub(1);
+        }
+        assert!(tests > 0)
```
Are you betting that at least one in a thousand a changed byte results in a valid PSET? :sweat_smile: Not sure why you'd want to enforce that? You could also just loop for anything between "both at least 1000 times and at least one test performed" or so?

RCasatta replied:
lol, I added during dev to ensure I was passing in the if branch and forgot to remove, removed in 682eb8a5bc972e28ccea8997ec582aa2d28769a1

RCasatta commented:
I think I am going to squash everything into 1 commit after review...

RCasatta commented on tests/pset.rs:
```diff
@@ -0,0 +1,335 @@
+        fn get_balances(&self) -> Value;
+    }
+
+    #[cfg_attr(feature = "integration", test)]
```
good point, removed in 682eb8a5bc972e28ccea8997ec582aa2d28769a1

RCasatta commented on tests/pset.rs:
```diff
@@ -0,0 +1,335 @@
+extern crate elementsd;
+extern crate rand;
+
+#[cfg(all(test, feature = "integration"))]
```
good idea, added `#![cfg...]` and removed the inner module

sanket1729 reviewed (commented):
Looks good to me. Left one small comment. Code review ACK. Can merge after squash

RCasatta commented:
added dd92f03896cfc4336e384b97f1df6223018b6816, 813ad20a3f972b179eb43b28a118d8b9e10c7544 and 657246ed736e2054ab0145f0d478aad5352ed10c

then I squashed into one

sanket1729 reviewed (approved):
code review ACK 31dd6bcb4f9a46b19abaf4d50626418fff5cc2bc

sanket1729 commented:
I don't know how to disable the 1.29 check in the CI. Will wait some time for @stevenroose's ACKs.

sanket1729 commented:
Was about to merge this. But it looks like I need admin access/not just write access. https://stackoverflow.com/questions/52589285/cannot-push-on-github-suddently
